opal/class/opal_fifo.h

```diff
@@ -194,7 +196,7 @@ static inline opal_list_item_t *opal_fifo_push_atomic (opal_fifo_t *fifo,
 
     opal_atomic_wmb ();
```

Comment:
Why we need this wmb ? there is nothing to flush, except the atomic swap right above.

Reply:
In fact I am not sure I understand how this code can be correct in the case where 2 threads are doing conflicting operations on a FIFO that contains only one element. Imagine a thread that pop while the second one push. The operation at line 202 (```tail_item->opal_list_next = item```) conflicts with the ```opal_atomic_sc_ptr``` in the pop, and can lead to a case where the head is NULL (it contains what the previous element assumed was his next) but the tail is not (it contains the result of the previous push).

Reply:
This scenario is tested by test/class/opal_fifo.c. The exhaustive test causes the threads to empty the fifo. 

The code that keeps  things consistent is at line 276. If the pop detected that the fifo was empty but something was being pushed (next == ghost && tail != item) it waits until the next pointer has been updated then sets the head to the new item. This is the same strategy used by vader and nemesis.

Reply:
Yeah, looking at it there is no way it can get to one of the following lines without the read and write completing anyway. That means the wmb does nothing. It shouldn't hurt though.

Reply:
If this is not critical path, it is fine ...